Dev sprint 16ver3

SleeplessShu opened the pull request:
Привет. Пожалуйста, если что-то нужно переделать, то я очень прошу, скажи как надо переделать, если заменить, то на что. Спасибо.

SleeplessShu committed:
Спринт 16.

Переработана активити Settings под MVVM

SleeplessShu committed:
Спринт 16.

Переработана активити Player под MVVM

SleeplessShu committed:
Спринт 16.

Переработана активити Search под MVVM

SleeplessShu committed:
Спринт 16.

Исправлен плеер
-лайвдата только между активити и вьюмоделью,
- состояния переключаются корректно

SleeplessShu committed:
Спринт 16.

Исправлен плеер
- утечки памяти при повторном воспроизведении

SleeplessShu committed:
Спринт 16.

внесены правки:
-плеер активити
-поиск активити
-лайвдата удалена из домена

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/search/ui/models/SearchViewState.kt:
```diff
@@ -0,0 +1,10 @@
+
+import com.practicum.playlistmaker_ver2.search.domain.models.Track
+
+data class SearchViewState(
+    val state: SearchState = SearchState.EMPTY,
+    val tracks: List<Track> = emptyList(),
+    val errorMessage: String? = null
+)
```
Удобнее моделировать стейт как через sealed interface/class

SleeplessShu replied:
исправил

SleeplessShu replied:
sealed не работает в init, а data работает

kekulta replied:
Не очень понял в чём именно твоя проблема. Можешь объяснить подробнее, что у тебя не получилось?

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/search/ui/ActivitySearch.kt:
```diff
@@ -0,0 +1,195 @@
+            hideKeyboard()
+        }
+
+        binding.bBackToMain.setOnClickListener(DebounceClickListener { finish() })
```
Отличная идея сделать DebounceClickListener!

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/settings/ui/SettingsViewModel.kt:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private val sharingInteractor = Creator.provideSharingInteractor()
+    private val settingsInteractor = Creator.provideSettingsInteractor()
```
Зависимости во ViewModel должны внедряться в параметрах, внутри ViewModelFactory.

SleeplessShu replied:
исправил

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/settings/ui/SettingsViewModel.kt:
```diff
@@ -0,0 +1,43 @@
+
+
+class SettingsViewModel : ViewModel() {
+    companion object {
+        fun getViewModelFactory(): ViewModelProvider.Factory = viewModelFactory {
+            initializer {
+                SettingsViewModel()
+            }
+        }
+    }
```
Согласно [код-стайлу](https://kotlinlang.org/docs/coding-conventions.html#class-layout) Котлина Companion Object должен находится в конце класса.

SleeplessShu replied:
исправил

SleeplessShu committed:
Спринт 16.

плеер, как сделать правильно?

SleeplessShu committed:
Спринт 16.

исправлен плеер и поиск

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/creator/Creator.kt:
Если для создания объекта в Creator тебе нужно инстансы какого-то класса, который так же предоставляет Creator, то тебе не обязательно прокидывать их в параметры.
Например, предположим, что для создания класс `Interactor` тебе нужно инстанс класса `Repository`.
Тогда ты можешь оформить это так:
```
fun getRepository(): Repository
fun getInteractor(rep: Repository): Interactor
```
Но в таком случае, в месте, где ты получаешь `Interactor` придётся делать следующую конструкцию:
```
Creator.getInteractor(Creator.getRepository)
```
Именно такое мы и видим у тебя в коде. Правильнее было бы сделать:
```
fun getRepository(): Repository
fun getInteractor(): Interactor { Creator.getRepository() /* Достаём репозиторий из креатора и создаём интерактор */ }
```

Затем, если окажется, что метод `getRepository()` не используется нигде вне креатора, то мы можем сделать его приватным, таким образом креатор будет предоставлять только то, что реально нужно его пользователям.

SleeplessShu replied:
вау, какой крутой совет. стало настолько лучше!!!
спасибо

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/search/ui/SearchViewModel.kt:
```diff
@@ -0,0 +1,139 @@
+class SearchViewModel(
+    private val searchInteractor: SearchInteractor, private val savedStateHandle: SavedStateHandle
+) : ViewModel() {
+    var currentQuery: String
+        get() = savedStateHandle.get<String>("currentQuery") ?: ""
+        set(value) {
+            savedStateHandle.set("currentQuery", value)
+        }
```
Не очень понятно зачем ты используешь здесь дополнительный хендл. Вьюмодель и так сохраняет стейт.

SleeplessShu replied:
1 - недостаток опыта и понимания, что именно я делаю. Пока что мои ощущения от разработки похожи на поиск камней на реке. Какие-то стукаешь и искорки, а какие-то ничего не делают.
2 - тут я пытался решить проблему с сохранением стейта, который не всегда сохранялся вьюмоделью)

SleeplessShu replied:
спасибо, что указал, я почитал и посмотрел и понял, что ты имеешь в виду и исправил

SleeplessShu committed:
Спринт 16.

Финальные правки после принятия спринта
-переработана передача аргументов в креатор
-переработан способ сохранения строки поиска
